Civi/Core/Url.php

```diff
@@ -0,0 +1,353 @@
+    }
+    return $this;
+  }
+
+  /**
+   * @return string|null
+   */
+  public function getQuery(): ?string {
+    return $this->query;
+  }
+
+  /**
+   * @param string|array|null $query
+   */
+  public function setQuery($query): Url {
+    $this->query = \CRM_Utils_System::makeQueryString($query);
+    return $this;
+  }
+
+  /**
+   * @param string|array $query
+   * @return $this
+   */
+  public function addQuery($query): Url {
+    if ($this->query === NULL) {
+      $this->query = \CRM_Utils_System::makeQueryString($query);
+    }
+    else {
+      $this->query .= '&' . \CRM_Utils_System::makeQueryString($query);
+    }
+    return $this;
+  }
+
+  /**
+   * @return string|null
+   */
+  public function getFragment(): ?string {
+    return $this->fragment;
+  }
+
+  /**
+   * @param string|null $fragment
+   */
+  public function setFragment(?string $fragment): Url {
+    $this->fragment = \CRM_Utils_System::makeQueryString($fragment);
+    return $this;
+  }
+
+  /**
+   * @param string|array $fragment
+   * @return $this
+   */
+  public function addFragment($fragment): Url {
+    if ($this->fragment === NULL) {
+      $this->fragment = \CRM_Utils_System::makeQueryString($fragment);
+    }
+    else {
+      $this->fragment .= '&' . \CRM_Utils_System::makeQueryString($fragment);
+    }
+    return $this;
+  }
+
+  /**
+   * @return string|null
+   *   'relative' or 'absolute'
+   */
+  public function getPreferFormat(): ?string {
+    return $this->preferFormat;
+  }
+
+  /**
+   * @param string|null $preferFormat
+   */
+  public function setPreferFormat(?string $preferFormat): Url {
+    $this->preferFormat = $preferFormat;
+    return $this;
+  }
+
+  /**
+   * @return bool
+   */
+  public function getHtmlEscape(): bool {
+    return $this->htmlEscape;
+  }
+
+  /**
+   * @param bool $htmlEscape
+   */
+  public function setHtmlEscape(bool $htmlEscape): Url {
+    $this->htmlEscape = $htmlEscape;
+    return $this;
+  }
+
+  /**
+   * @return bool|null
+   */
+  public function getSsl(): ?bool {
+    return $this->ssl;
+  }
+
+  /**
+   * @param bool|null $ssl
+   */
+  public function setSsl(?bool $ssl): Url {
+    $this->ssl = $ssl;
+    return $this;
+  }
+
+  /**
+   * @param string $flags
+   *   A series of flag-letters. Any of the following:
+   *   - [a]bsolute
+   *   - [r]elative
+   *   - [h]tml
+   *   - [s]sl
+   * @return $this
+   */
+  public function useFlags(string $flags): Url {
+    $len = strlen($flags);
+    for ($i = 0; $i < $len; $i++) {
+      switch ($flags[$i]) {
+        // (a)bsolute url
+        case 'a':
+          $this->preferFormat = 'absolute';
+          break;
+
+        // (r)elative url
+        case 'r':
+          $this->preferFormat = 'relative';
+          break;
+
+        // (h)tml encoding
+        case 'h':
+          $this->htmlEscape = TRUE;
+          break;
+
+        // (p)lain text encoding
+        case 'p':
+          $this->htmlEscape = FALSE;
```

Comment:
Ooh, an undocumented flag that overrules `htmlize`! Yet another blow to its fragile ego... perhaps leading it to the brink of a well-deserved existential crisis.

Reply:
@totten this flag is still undocumented. Should it be removed?

Reply:
The "undocumented" note had confused me, but I see now: it's documented on `Civi::url(...$flags...)` but not `Url::useFlags($flags)`. That's easy to cleanup.

I should take a stab at the Smarty side before deciding whether to remove that flag. My suspicion is that it'll make it easier to tune Smarty. (Quick estimate: in core, there are ~530 places where Smarty wants HTML and ~130 places where it wants plain text.)